### PR TITLE
[compiler] Add a Loc to Param and TypeParam in Move Model

### PR DESCRIPTION
**Pull request by brmataptos**
### Description

Fix Issue #10731 by adding a `Loc` to `Param` and `TypeParam` in `move-model` to more accurately provide errors about parameters.

### Test Plan

Looked at some cases where this is illustrated.  Still need an error case where there is a failure *after* inlining.

### CODE REVIEW

File: third_party/move/evm/move-to-yul/src/abi_native_functions.rs [hunk 1]
```diff
@@ -294,7 +294,7 @@ impl SoliditySignature {
     pub(crate) fn create_default_sig_for_encode_fun(ctx: &Context, fun: &FunctionEnv<'_>) -> Self {
         let fun_name = fun.symbol_pool().string(fun.get_name()).to_string();
         let mut para_type_lst = vec![];
-        for Parameter(para_name, move_ty) in fun.get_parameters() {
+        for Parameter(para_name, move_ty, _) in fun.get_parameters() {
```
Comment (vineethk) [inline third_party/move/evm/move-to-yul/src/abi_native_functions.rs:297]: Not an action item for this PR, but we should ask around if this crate is being maintained and is actually usable.

If not, we might be better off getting rid of it sometime soon.

File: third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs [hunk 2]
```diff
@@ -372,15 +372,15 @@ impl ModuleGenerator {
         let type_parameters = fun_env
             .get_type_parameters()
             .into_iter()
-            .map(|TypeParameter(_, TypeParameterKind { abilities, .. })| abilities)
+            .map(|TypeParameter(_, TypeParameterKind { abilities, .. }, _loc)| abilities)
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs:375]:
```suggestion
            .map(|TypeParameter(_, TypeParameterKind { abilities, .. }, _)| abilities)
```

Comment (vineethk) [reply third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs:375]: nit: it is a bit weird to name one of your ignores, but not the other.

File: third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs [hunk 3]
```diff
@@ -372,15 +372,15 @@ impl ModuleGenerator {
...
             .collect::<Vec<_>>();
         let parameters = self.signature(
             ctx,
             loc,
             fun_env
                 .get_parameters()
                 .iter()
-                .map(|Parameter(_, ty)| ty.to_owned())
+                .map(|Parameter(_, ty, _loc)| ty.to_owned())
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs:383]:
```suggestion
                .map(|Parameter(_, ty, _)| ty.to_owned())
```

File: third_party/move/move-compiler-v2/src/bytecode_generator.rs [hunk 4]
```diff
@@ -785,7 +785,7 @@ impl<'env> Generator<'env> {
             .get_function(fun)
             .get_parameters()
             .into_iter()
-            .map(|Parameter(_, ty)| ty.instantiate(&type_args))
+            .map(|Parameter(_, ty, _loc)| ty.instantiate(&type_args))
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/bytecode_generator.rs:788]:
```suggestion
            .map(|Parameter(_, ty, _)| ty.instantiate(&type_args))
```

File: third_party/move/move-compiler-v2/src/inliner.rs [hunk 5]
```diff
@@ -805,14 +805,8 @@ impl<'env, 'rewriter> InlinedRewriter<'env, 'rewriter> {
...
             .map(|param| {
-                let Parameter(sym, ty) = *param;
-                // TODO(10731): ideally, each Parameter has its own loc.  For now, we use the
-                // function location.  body should have types rewritten, other inlining complete,
-                // lambdas inlined, etc.
-                let id = env.new_node(
-                    function_loc.clone().inlined_from(call_site_loc),
-                    ty.instantiate(self.type_args),
-                );
+                let Parameter(sym, ty, loc) = *param;
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/inliner.rs:808]: Nice!

File: third_party/move/move-model/src/builder/exp_builder.rs [hunk 6]
```diff
@@ -424,7 +425,7 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
     fn type_display_context(&self) -> TypeDisplayContext<'_> {
         TypeDisplayContext {
             env: self.parent.parent.env,
-            type_param_names: Some(self.type_params.iter().map(|(s, _)| *s).collect()),
+            type_param_names: Some(self.type_params.iter().map(|(s, _, _loc)| *s).collect()),
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/exp_builder.rs:428]:
```suggestion
            type_param_names: Some(self.type_params.iter().map(|(s, _, _)| *s).collect()),
```

File: third_party/move/move-model/src/builder/exp_builder.rs [hunk 7]
```diff
@@ -2419,11 +2441,13 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
                     Operation::Select(mid, sid, FieldId::new(field_name))
                 } else {
                     // Create a placeholder for later resolution.
-                    self.placeholder_map
-                        .insert(id, ExpPlaceholder::FieldSelectInfo {
+                    self.placeholder_map.insert(
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/exp_builder.rs:2444]: Not sure if this is worth investigating, but this file seems to have a number of purely formatting related changes (like this one).

File: third_party/move/move-model/src/builder/module_builder.rs [hunk 8]
```diff
@@ -1353,11 +1354,11 @@ impl<'env, 'translator> ModuleBuilder<'env, 'translator> {
                     et.set_translate_move_fun()
                 }
                 let loc = et.to_loc(&body.loc);
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/module_builder.rs:1356]: Move this closer to its one use down a few lines?

File: third_party/move/move-model/src/builder/module_builder.rs [hunk 9]
```diff
@@ -1441,7 +1442,7 @@ impl<'env, 'translator> ModuleBuilder<'env, 'translator> {
             let no_mut_ref_param = self.spec_funs[spec_fun_idx]
                 .params
                 .iter()
-                .map(|Parameter(_, ty)| !ty.is_mutable_reference())
+                .map(|Parameter(_, ty, _loc)| !ty.is_mutable_reference())
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/module_builder.rs:1445]:
```suggestion
                .map(|Parameter(_, ty, _)| !ty.is_mutable_reference())
```

File: third_party/move/move-prover/tests/sources/functional/schema_name.exp [hunk 10]
```diff
@@ -1,6 +1,9 @@
 Move prover returns: exiting with model building errors
-error: Variable `c` conflicts with a specification variable in the schema TestSchemaName::TestNameConflict
-  ┌─ tests/sources/functional/schema_name.move:9:16
-  │
-9 │             x: c + 3
-  │                ^^^^^
+error: A specification variable in the schema TestSchemaName::TestNameConflict conflicts with
```
Comment (vineethk) [inline third_party/move/move-prover/tests/sources/functional/schema_name.exp:2]: Looked like this sentence ended abruptly. It is not immediately clear that I am supposed to continue reading to the last line.

File: third_party/move/move-prover/move-docgen/src/docgen.rs [hunk 11]
```diff
@@ -1577,7 +1579,7 @@ impl<'env> Docgen<'env> {
         let type_param_names = func_env
             .get_type_parameters()
             .iter()
-            .map(|TypeParameter(name, _)| *name)
+            .map(|TypeParameter(name, _, _loc)| *name)
```
Comment (vineethk) [inline third_party/move/move-prover/move-docgen/src/docgen.rs:1582]:
```suggestion
            .map(|TypeParameter(name, _, _)| *name)
```

File: third_party/move/move-prover/move-docgen/src/docgen.rs [hunk 12]
```diff
@@ -1590,7 +1592,7 @@ impl<'env> Docgen<'env> {
         let type_param_names = struct_env
             .get_type_parameters()
             .iter()
-            .map(|TypeParameter(name, _)| *name)
+            .map(|TypeParameter(name, _, _loc)| *name)
```
Comment (vineethk) [inline third_party/move/move-prover/move-docgen/src/docgen.rs:1595]:
```suggestion
            .map(|TypeParameter(name, _, _)| *name)
```

File: third_party/move/move-prover/bytecode-pipeline/src/number_operation_analysis.rs [hunk 13]
```diff
@@ -353,9 +353,10 @@ impl<'a> NumberOperationAnalysis<'a> {
                                                     // Still need to check compatibility
                                                     if !allow_merge && oper.conflict(&para_vec[i]) {
                                                         self.func_target.global_env().error(
-                                                            &self
-                                                                .func_target
-                                                                .get_bytecode_loc(attr_id),
+                                                            loc,
+                                                            // &self
```
Comment (vineethk) [inline third_party/move/move-prover/bytecode-pipeline/src/number_operation_analysis.rs:357]: Leftover?

File: third_party/move/move-prover/bytecode-pipeline/src/mono_analysis.rs [hunk 14]
```diff
@@ -456,7 +456,7 @@ impl<'a> Analyzer<'a> {
     fn analyze_spec_fun(&mut self, fun: QualifiedId<SpecFunId>) {
         let module_env = self.env.get_module(fun.module_id);
         let decl = module_env.get_spec_fun(fun.id);
-        for Parameter(_, ty) in &decl.params {
+        for Parameter(_, ty, _loc) in &decl.params {
```
Comment (vineethk) [inline third_party/move/move-prover/bytecode-pipeline/src/mono_analysis.rs:459]:
```suggestion
        for Parameter(_, ty, _) in &decl.params {
```

File: third_party/move/move-prover/boogie-backend/src/spec_translator.rs [hunk 15]
```diff
@@ -376,7 +376,9 @@ impl<'env> SpecTranslator<'env> {
                 boogie_name,
                 fun.params
                     .iter()
-                    .map(|Parameter(n, _)| { format!("{}", n.display(module_env.symbol_pool())) })
+                    .map(|Parameter(n, _, _loc)| {
```
Comment (vineethk) [inline third_party/move/move-prover/boogie-backend/src/spec_translator.rs:379]:
```suggestion
                    .map(|Parameter(n, ..)| {
```

**Review by vineethk — APPROVED** [review]
I only have minor comments, LGTM otherwise.

File: third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs (hunk 2 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs:375]: Well, everyone knows the first field is always `NodeId`.  But what is the 3d one?

File: third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs (hunk 3 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs:383]: Done.

File: third_party/move/move-compiler-v2/src/bytecode_generator.rs (hunk 4 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/src/bytecode_generator.rs:788]: Done

File: third_party/move/move-model/src/builder/exp_builder.rs (hunk 6 above)
Comment (brmataptos) [reply third_party/move/move-model/src/builder/exp_builder.rs:428]: Done.

File: third_party/move/move-model/src/builder/module_builder.rs (hunk 9 above)
Comment (brmataptos) [reply third_party/move/move-model/src/builder/module_builder.rs:1445]: Done.

File: third_party/move/move-prover/move-docgen/src/docgen.rs (hunk 11 above)
Comment (brmataptos) [reply third_party/move/move-prover/move-docgen/src/docgen.rs:1582]: Done

File: third_party/move/move-prover/boogie-backend/src/spec_translator.rs (hunk 15 above)
Comment (brmataptos) [reply third_party/move/move-prover/boogie-backend/src/spec_translator.rs:379]: Done.

File: third_party/move/move-prover/bytecode-pipeline/src/mono_analysis.rs (hunk 14 above)
Comment (brmataptos) [reply third_party/move/move-prover/bytecode-pipeline/src/mono_analysis.rs:459]: Done

File: third_party/move/move-prover/bytecode-pipeline/src/number_operation_analysis.rs (hunk 13 above)
Comment (brmataptos) [reply third_party/move/move-prover/bytecode-pipeline/src/number_operation_analysis.rs:357]: Done.

File: third_party/move/move-prover/move-docgen/src/docgen.rs (hunk 12 above)
Comment (brmataptos) [reply third_party/move/move-prover/move-docgen/src/docgen.rs:1595]: Done.

File: third_party/move/move-model/src/builder/module_builder.rs (hunk 8 above)
Comment (brmataptos) [reply third_party/move/move-model/src/builder/module_builder.rs:1356]: This seems to be just outside the 3-line loop that uses it.

File: third_party/move/move-prover/tests/sources/functional/schema_name.exp (hunk 10 above)
Comment (brmataptos) [reply third_party/move/move-prover/tests/sources/functional/schema_name.exp:2]: I tried to fix this code, but i'm not sure how to make this file regenerate.

File: third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs (hunk 2 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs]: Well, "everyone" is incorrect: the first field here is symbol, not node id.

File: third_party/move/move-prover/tests/sources/functional/schema_name.exp (hunk 10 above)
Comment (vineethk) [reply third_party/move/move-prover/tests/sources/functional/schema_name.exp:2]: Someone with prover experience could help? @rahxephon89

File: third_party/move/move-model/src/builder/module_builder.rs (hunk 9 above)
Comment (brmataptos) [reply third_party/move/move-model/src/builder/module_builder.rs]: I finally got it to update.  For some reason, I seem to have to run "UB=1 git test -p move-prover" in isolation to update the file.

Comment (vineethk) [reply third_party/move/move-model/src/builder/module_builder.rs]: What is `git test`? Or perhaps you meant `cargo test`?

File: third_party/move/move-model/src/builder/exp_builder.rs (hunk 7 above)
Comment (brmataptos) [reply third_party/move/move-model/src/builder/exp_builder.rs]: I see what you mean, but `rustfmt +nightly` seems to be fine with it.  I think that changing `loc` to `_loc` in a few places put the line length over some limit.

**Review by rahxephon89 — APPROVED** [review]

**Pull request merged**